fdbserver/BlobManager.actor.cpp

```diff
@@ -1314,6 +1320,585 @@ ACTOR Future<Void> blobWorkerRecruiter(
+									wait(dummy->onError(e));
+								}
+							}
+						}
+						break;
+					}
+
+					// otherwise, there are no changes and we should wait until the next change (or timeout)
+					state Future<Void> watchPruneIntentsChange = tr->watch(blobGranulePruneChangeKey);
+					wait(tr->commit());
+
+					if (BM_DEBUG) {
+						printf("monitorPruneKeys waiting for change or timeout\n");
+					}
+
+					choose {
+						when(wait(watchPruneIntentsChange)) {
+							if (BM_DEBUG) {
+								printf("monitorPruneKeys saw a change\n");
+							}
+							tr->reset();
+						}
+						when(wait(delay(SERVER_KNOBS->BG_PRUNE_TIMEOUT))) {
+							if (BM_DEBUG) {
+								printf("monitorPruneKeys got a timeout\n");
+							}
+							break;
+						}
+					}
+				} catch (Error& e) {
+					wait(tr->onError(e));
+				}
+			}
+
+			tr->reset();
+
+			if (BM_DEBUG) {
+				printf("Looping over prune intents\n");
+			}
+
+			// loop through all prune intentions and do prune work accordingly
+			state KeyRef beginKey = normalKeys.begin;
+			loop {
+				tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+				tr->setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+
+				state std::vector<Future<Void>> prunes;
+				try {
+					// TODO: replace 10000 with a knob
+					KeyRange nextRange(KeyRangeRef(beginKey, normalKeys.end));
+					state RangeResult pruneIntents = wait(krmGetRanges(
+					    tr, blobGranulePruneKeys.begin, nextRange, 10000, GetRangeLimits::BYTE_LIMIT_UNLIMITED));
+
+					// TODO: would we miss a range [pruneIntents[9999], pruneIntents[10000]) because of the `more`?
+					//       Or does `readThrough` take care of this? We also do this in recoverBlobManager
+					printf("pruneIntents.size()==%d\n", pruneIntents.size());
+					for (int rangeIdx = 0; rangeIdx < pruneIntents.size() - 1; ++rangeIdx) {
+						if (pruneIntents[rangeIdx].value.size() == 0) {
+							continue;
+						}
+						KeyRef rangeStartKey = pruneIntents[rangeIdx].key;
+						KeyRef rangeEndKey = pruneIntents[rangeIdx + 1].key;
+						KeyRange range(KeyRangeRef(rangeStartKey, rangeEndKey));
+						Version pruneVersion;
+						bool force;
+						std::tie(pruneVersion, force) = decodeBlobGranulePruneValue(pruneIntents[rangeIdx].value);
+
+						printf("about to prune range [%s-%s) @ %d, force=%s\n",
+						       rangeStartKey.printable().c_str(),
+						       rangeEndKey.printable().c_str(),
+						       pruneVersion,
+						       force ? "T" : "F");
+						prunes.emplace_back(pruneRange(self, rangeStartKey, rangeEndKey, pruneVersion, force));
+					}
+
+					// wait for this set of prunes to complete before starting the next ones since if we prune
+					// a range R at version V and while we are doing that, the time expires, we will end up
+					// trying to prune the same range again since the work isn't finished and the prunes will race
+					//
+					// TODO: this isn't that efficient though. Instead we could keep metadata as part of the BM's
```

Comment:
This is a performance optimization that I think we can address in a future iteration

Reply:
I also decided _not_ to clear the prune intent key here (as opposed to in `pruneRange` itself) since I thought it might be a decent amount of work to redo all the prunes if just one failed. Doing it in `pruneRange` allowed me to evaluate success on a range-by-range basis.